### ISSUE TITLE
Schemas Proposals

### ISSUE THREAD

**Issue by cliffconklin** [zero-to-mastery/projectLive#21]
// Proposals
//  Models:
//    For simplicity and reduction of required queries, Embedded Data 
//    Models should be used. 
// 
//    *** After working some of these documents, if there is not an issue with
//        database performance, I see a lot of common pieces of data that
//        could be shared and maybe should be their own document.   Thoughts??
//
//  Messages:
//    Using MongoDB’s capped collections and tailable cursors we can 
//    build a simple pub/sub system to communicate messages (documents) 
//    between processes.  There may be an issue here however with also needing
//    some sort of pub/sub service (something like Pusher, www.pusher.com)
//      
//  Chat:
//    For messages between users (if we need that), user, and message 
//    information will be used
//
//   
// User Document
{
    _id: <ObjectId>
    user: "<123abc>"
    pwd: <"<password>"
    contact: {
               phone: "123-456-7890",
               email: "xyz@example.com"
    },
    oauth: {
             token : "tokenId",
             provider: "oauthProvider??"
    },
    access: {
             level: 2,
             group: "admin"
    }
}
// Field           Type        Description
// ----------------------------------------------------------------------------
// user            string      Name of the user
// pwd             string      The user's password
// contact         embedded    Contact information for the user
// oauth           embedded    OAuth information
// access          embedded    access priviledges for the user
// User Message Document  (maybe this could be put in the User Document 
// above as an Embedded Data Model ??)
{
    _id: <ObjectId>
    user: "123abc"
    logs: { 
            last_login: 11.03.2009, 
            last_login_reset: 04.01.2003
    },
    state: {
             online: true, 
             available: false
    }
}
// Conversation
{
    _id: <ObjectId>
    members: [user_id1, user_id2
    ],
    messages: [
        { author: user_2, body: 'Hi, how are you'
        },
        { author: user_1, body: 'Good, nice to see you  :)'
        },
        { author: user_2, body: 'Nice day here, later...'
        },
        { author: user_1, body: 'Good day'
        }
    ]
}
// Stream
Stream Document
{
    _id: <ObjectId>,
    stream: {
        id: stream_id,
        live: live_stream,
        replay: replay_stream
    }
}
// Statistics
Counter Document
{
    _id: <ObjectId>
    counter: {
        id: stream_id,
        viewers: live_viewers,
        views total_views
    }
}

**Comment by matfire** [comment]
Thanks for taking on this task: I've looked at the schemas above, and I see your points about embedding the messages inside the user model. However, since this data will also be used on a dashboard, it may be irrevelant too query all the users just to get the messages for a stream; we should probably check whether the time required to do a relational query outweights the overhead of a complicated query with tons of messages

**Comment by cliffconklin** [comment]
Makes sense.  I guess we could have a chat on discord maybe to clarify the needs of these schemas in some more detail, no?
